### PR TITLE
Add object information to MooseObject::moose[Error/Warning]()

### LINKED ISSUE #16863: Add useful object information to MooseObject::moose{Error/Warning}()
## Reason
We know things like the object type and name when calling `MooseObject::moose{Error/Warning}()`. In objects that aren't created using complicated actions, we even have the input and line number where the object is defined. Why not use this information?

## Design
Append something like "<object type> "<object name>" <possible input location for object> to the beginning of `MooseObject::moose{Error/Warning}()`.

## Impact
More detailed errors for MooseObject.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #16863

There is one issue with this: for errors and warnings that already include some of this information, we will get repetetive information. However, for warnings and errors that _do not_ have extraneous information, this adds a LOT of useful information to the error. I think it's better to have extraneous information than no information at all.

### CODE REVIEW

**Comment by loganharbour** [comment]
This is particularly useful for when we call `mooseError()` for parameters that are _missing_.

For example:
```
*** ERROR ***
enameBoundaryGenerator "rename" (/Users/harblh/projects/moose/test/tests/meshgenerators/rename_boundary_generator/rename_boundary.i:10):

RenameBoundaryGenerator: You must supply exactly one of old_boundary_id or old_boundary_name
```

**Comment by aeslaughter** [comment]
I think this information is useful, but should think about how it is presented. In the message above it wasn't exactly clear what was being reported. I think it should be something a bit more descriptive:

```
*** ERROR ***
/Users/harblh/projects/moose/test/tests/meshgenerators/rename_boundary_generator/rename_boundary.i:10
The error occurred in the class RenameBoundaryGenerator with name "rename"

You must supply exactly one of old_boundary_id or old_boundary_name
```

Also, could/should we include source information? If this is a MooseObject then I think the file in which the class is registered, is available.

```
*** ERROR ***
/Users/harblh/projects/moose/test/tests/meshgenerators/rename_boundary_generator/rename_boundary.i:10
The error occurred in the class RenameBoundaryGenerator with name "rename"
This object is registered in:
/Users/.../RenameBoundaryGenerator.C

You must supply exactly one of old_boundary_id or old_boundary_name
```

I am all for more information, we can work on improving the format. Perhaps even use some color to help highlight the various information.

**Comment by loganharbour** [comment]
> I am all for more information, we can work on improving the format. Perhaps even use some color to help highlight the various information.

Yeah - I agree. Let's discuss with a few tomorrow.

**Comment by loganharbour** [comment]
After a conversation with @aeslaughter and @friedmud, we've agreed on the following:

## Error/warning with input file available

```
<input:line>:
The <error/warning> occurred in the object <type()> "<name()>".

<user error>
```

Looks like

```
*** ERROR ***
/Users/harblh/projects/moose/test/tests/meshgenerators/rename_boundary_generator/rename_boundary.i:10:
The error occurred in the object RenameBoundaryGenerator named "rename".

You must supply exactly one of old_boundary_id or old_boundary_name
```

## Error without input file available

```
The <error/warning> occurred in the object <type()> "<name()>".

<user error>
```

@idaholab/moose-team I'd like some input on this, if possible. It's a very global change.

**Comment by joshuahansel** [comment]
- I'm all for the automatic inclusion of object name/class information in error messages. The duplication issue you mentioned is going to be a bit ugly but at least does not decrease information. Application developers can always fix their error messages to avoid this, and that should be their responsibility if it bothers them.
- One thing I'm unsure about is how you've presented the line number. Presenting the line number without context makes it seem like a parsing error, rather than something occurring after parsing.
- I suggest rewording "in the object <type>..." since you're actually giving the object type, not the object name first.

All in all here's my suggestion:
```
The error below occurred in the object "rename", of type "RenameBoundaryGenerator", created on the following line:
/Users/harblh/projects/moose/test/tests/meshgenerators/rename_boundary_generator/rename_boundary.i:10

You must supply exactly one of old_boundary_id or old_boundary_name
```
or
```
/Users/harblh/projects/moose/test/tests/meshgenerators/rename_boundary_generator/rename_boundary.i:10:
The following error occurred in the object "rename", of type "RenameBoundaryGenerator", created on the line above:

You must supply exactly one of old_boundary_id or old_boundary_name
```
Or just ignore my line number comments because it's not a huge deal either way :)

**Comment by rwcarlsen** [comment]
I don't think I really like this.  Adding more context to errors is good, but there are many errors that occur not necessarily because there was any problem with the object where the error is emitted. Also, if the error is related to config for the object from the input file, paramError is almost always what you want. For rare occasions where we have some error that is user-caused from an input file but isn't related to a parameter, something like objectError maybe?  But I think paramError should really cover the vast majority of cases where these modifications actually would help.

**Comment by loganharbour** [comment]
After some thought and the discussion above, I think it's best to just have:

```
The following error occurred in the object "rename", of type "RenameBoundaryGenerator".

You must supply exactly one of old_boundary_id or old_boundary_name
````

**Review by aeslaughter — DISMISSED** [review]

**Review by aeslaughter — APPROVED** [review]
